### ISSUE TITLE
Support Arabic and RTL in Tensorflow documentation

### ISSUE THREAD

**Issue by mohamed-ali** [tensorflow/docs-l10n#129]
As suggested by @lamberta, I am creating this issue to gather information about the extent to which Tensorflow website supports Arabic and RTL (Right to left) layout in general. 

In fact, I started translating two of Tensorflow's tutorials into Arabic with the following two PRs:
* https://github.com/tensorflow/docs-l10n/pull/121
* https://github.com/tensorflow/docs-l10n/pull/122

Knowing that Google Colab, like jupyter notebooks, supports html tags, I used them to ensure that the RTL text is displayed correctly, while keeping the LTR (left to right) layout for everything else in the tutorial, since the code needs to be LTR. Which results in a correct page layout as you can see in the following two notebooks rendering from the above two PRs:

* https://colab.research.google.com/github/abjed/docs-l10n/blob/ar-quickstart/site/ar/tutorials/quickstart/beginner.ipynb
* https://colab.research.google.com/github/abjed/docs-l10n/blob/ar-quickstart-advanced/site/ar/tutorials/quickstart/advanced.ipynb

Opening these notebooks in https://nbviewer.jupyter.org/ , which renders the notebook to html,  shows that the RTL layout defined by wrapping the Arabic text with `<div dir="rtl"></div>` works as intended. As you can see in the following link:

* https://nbviewer.jupyter.org/github/abjed/docs-l10n/blob/ar-quickstart/site/ar/tutorials/quickstart/beginner.ipynb

So, my questions, are: 

1. Does the pipeline that generates the tutorials from the notebooks preserves these HTML tags, and displays them correctly (we might need to do an experiment)?
2. If not is there an alternative to support Arabic (RTL) properly? 

Thanks.

**Comment by lamberta** [comment]
Thanks @mohamed-ali 

Just making a note to test wrapping Markdown text within a `<div>` since I think this doesn't work in some Markdown parsers (for the site?)

**Comment by mohamed-ali** [comment]
@lamberta I understand that wrapping the markdown with `<div dir="rtl"></div>` could disable the markdown parser on the wrapped text. However, for a first version, having an understandable/readable content with the correct display is more important than minor styling. Also, as an arabic reader, a text mixing RTL and LTR words, if not properly displayed loses all meaning. 

In the other hand, most of stylings impacted with the `<div dir="rtl"></div>` workaround can be replaced with standard html alternatives very easily, here are some examples:

Links
```
[Tensorflow website](https://www.tensorflow.org/)
```
is equivalent to
```html
<a href="https://www.tensorflow.org/">Tensorflow website</a>
```

Code
```
`tf.Keras`
```
is equivalent to
```
<code>tf.Keras</code>
```
etc. 

So I suggest we move forward and do an initial experiment with these html workarounds. 
We can always improve iteratively in the future.

@lamberta  What do you think?

**Comment by lamberta** [comment]
> having an understandable/readable content with the correct display is more important than minor styling.

It will break elements like lists. Also, among our preprocessing steps, we use backticks around code symbols to auto-link to the corresponding API reference page.

I support adding right-to-left support, but we need to investigate where in the pipeline this should be added. Requiring Arabic community contributors to know when to use this particular element vs something else is too much---which will lead to more bugs filed when pages don't render as intended.

> I suggest we move forward and do an initial experiment with these html workarounds. We can always improve iteratively in the future.

I agree that we will need to iterate. Unfortunately, due to the COVID-19 situation, we are working on this feature request as time allows. See Martin's [message](https://groups.google.com/a/tensorflow.org/d/msg/docs/BdZpILm3__g/9FBJcONfAwAJ). Thanks

**Comment by mohamed-ali** [comment]
@lamberta  It seems that there is an easier alternative to achieve correct RTL display: The layout can be fixed as a post-processing using the following CSS. 

```css
p, ul, ol, h1, h2, h3, h4, h5, h6 {
  direction: rtl;
}
``` 

This has the added advantage of keeping the current pipeline with all its features, that you described, intact.

I tried these rules on the live website https://www.tensorflow.org/tutorials/keras/classification and it changes the relevant page sections to RTL properly.

<img width="512" alt="Screenshot 2020-03-18 at 18 21 33" src="https://user-images.githubusercontent.com/2883926/76989272-6026bc80-6946-11ea-9476-e28d02f996cd.png">
<img width="518" alt="Screenshot 2020-03-18 at 18 21 46" src="https://user-images.githubusercontent.com/2883926/76989276-6157e980-6946-11ea-9dac-3cbf2dda5e78.png">
<img width="522" alt="Screenshot 2020-03-18 at 18 21 59" src="https://user-images.githubusercontent.com/2883926/76989284-6452da00-6946-11ea-843e-0894ce5d0b29.png">

If it's possible to add a custom css file per language then the problem should be easy to solve.

**Comment by lamberta** [comment]
Ah, that is interesting. Good one
I've filed an internal issue (b/151835181) to communicate with the site infra and Colab teams to see if we can decide on the "right" solution for this.

If we add the CSS for the site, do any browser display settings kick in on COlab if it detects a locale or anything?

**Comment by mohamed-ali** [comment]
If we want to, it is possible to add a logic that loads the rtl file when the user is accessing the website from a country that requires RTL. 

I also checked on colab, the same css rules as above work
```css
p, ul, ol, h1, h2, h3, h4, h5, h6 {
  direction: rtl;
}
```
Here is the original link https://colab.research.google.com/notebooks/intro.ipynb, and here's how it looks with the css corrections:

<img width="707" alt="Screenshot 2020-03-18 at 18 45 03" src="https://user-images.githubusercontent.com/2883926/76991135-623e4a80-6949-11ea-9531-bd4e436c771b.png">
<img width="705" alt="Screenshot 2020-03-18 at 18 45 19" src="https://user-images.githubusercontent.com/2883926/76991137-636f7780-6949-11ea-93ab-037d42a31776.png">

To help colab load the css rtl, we can send a hint  as url parameter in the GET request , for example:

https://colab.research.google.com/notebooks/intro.ipynb&dir=rtl 

Any other way to detect the language or locale of the text can also be used to decide whether or not to load the RTL css file.

**Comment by lamberta** [comment]
Ok, the site infra for tensorflow.org should already support this (though we need to test).
Colab does not support RTL but there is a feature request (b/110534608) and I'll bump that.

So, for now, let's add a notebook *without* `rtl` divs or CSS to see where we're at. And then work on some of the underlying infra issues before adding workarounds (which may complicate our docs pipeline).

**Comment by lamberta** [comment]
Ok, new plan ...

At this time, there are no plans for Colab to support the "toggle rtl layout" preference that Jupyter currently has. So:

1. `ar/` notebooks will need to wrap all *text* cells in `<div dir="rtl">`. This seems to render the inner Markdown and HTML content correctly in both Colab and Jupyter.
2. We need to add support to our docs pipeline to strip these divs before building the HTML for the site. The site should currently work as-is.

Will not be using CSS or URL params.

**Comment by mohamed-ali** [comment]
That's great. 

Few clarifications about the notebooks linked above which are rendered from the two PRs I put in the issue description:

Links don't seem to be rendered correctly:
```
[something](https://somelink.com) 
```
Therefore, I wrapped them with
```
<a href="https://somelink.com">something</a>
```

Code doesn't render 
```
`tf.Keras`
```
Therefore, I used <code>
```
<code>tf.Keras</code>
```

The `code` isn't  mandatory, it  can be kept as is. However rendering links correctly is useful.

Potential solution:

1. Strength colab's ability to render these tags even within `<div dir="rtl"></div>`
2. I can push a PR to the pipeline to convert the tags html back to their markdown counterparts before preparing the website. They can remains in the notebooks for correct rendering.

**Comment by mohamed-ali** [comment]
We can use this PR to test the above plan: https://github.com/tensorflow/docs-l10n/pull/121

**Comment by lamberta** [comment]
My tests in Colab show the following renders correctly in a text cell. Are you seeing something different? ...

```
<div dir="rtl">

* Foo
* [Bar](http://localhost)
* `tf.keras`

</div>
```

As mentioned, among our preprocessing steps, we use backticks around code symbols to auto-link to the corresponding API reference page. Aside from adding the wrapping `<div dir="rtl">` formatting should remain consistent with the source-of-truth. Since these notebook are rendered in a few places, the priorities are: 1. tensorflow.org, 2. Colab, 3. Jupyter, 4. GitHub preview.

**Comment by mohamed-ali** [comment]
@lamberta You are right, what your example renders correctly

```
<div dir="rtl">

* Foo
* [Bar](http://localhost)
* `tf.keras`

</div>
```

However, it doesn't render correctly without the whitespace, as follows:

```
<div dir="rtl">
* Foo
* [Bar](http://localhost)
* `tf.keras`
</div>
```

**Comment by mohamed-ali** [comment]
@lamberta  I updated the PR https://github.com/tensorflow/docs-l10n/pull/121 to remove all the use of html except `<div dir="rtl"></div>`. Can we use it to test the output?

**Comment by lamberta** [comment]
Yes, looks good. Thanks
I tagged in @baheytharwat to review that notebook. Once merged, we can use that to test. We still need to add functionality on our side, though.

**Comment by yashk2810** [comment]
This is now live: https://www.tensorflow.org/tutorials/quickstart/beginner?hl=ar

**Comment by lamberta** [comment]
Great! Thanks, Yash.
And congrats to @mohamed-ali , @baheytharwat , and @tayciryahmed . I look forward to seeing the community grow

**Comment by mohamed-ali** [comment]
It's great to see the results. Thanks @lamberta and @yashk2810 for your assistance in bringing this to life. 

Thanks @baheytharwat  and @tayciryahmed

**Comment by yashk2810** [comment]
No problem :)

Keep those PR's coming!
